File: python/plumr/stream_text.py
"""High-level callback-style helpers that mirror AI-SDK's `streamText`."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, List, Mapping, Optional

from .client import AsyncPlumr, Plumr
from .types import (
    ErrorEvent,
    LlmDeltaEvent,
    PlumrEvent,
    ReasoningDeltaEvent,
    RunEndEvent,
    RunInput,
    ToolCallEvent,
)


@dataclass
class StreamTextResult:
    text: str
    reasoning: str
    runId: str
    status: str
    error: Optional[str]
    durationMs: int
    conversationId: Optional[str] = None
    totalCostUsd: Optional[float] = None
    totalPromptTokens: Optional[int] = None
    totalCompletionTokens: Optional[int] = None
    toolCalls: List[ToolCallEvent] = field(default_factory=list)
    errors: List[ErrorEvent] = field(default_factory=list)


def stream_text(
    client: Plumr,
    input: RunInput,
    *,
    params: Optional[Mapping[str, Any]] = None,
    conversation_id: Optional[str] = None,
    idempotency_key: Optional[str] = None,
    on_text: Optional[Callable[[str, str], None]] = None,
    on_reasoning: Optional[Callable[[str, str], None]] = None,
    on_tool_call: Optional[Callable[[ToolCallEvent], None]] = None,
    on_error: Optional[Callable[[ErrorEvent], None]] = None,
    on_event: Optional[Callable[[PlumrEvent], None]] = None,
) -> StreamTextResult:
    """Stream a run with callbacks; return aggregate text and metadata.

    Each `on_*` callback receives the relevant payload. `on_text` and
    `on_reasoning` receive `(text_chunk, node_id)`. `on_event` is called
    for every event (after the typed callback fires) as an escape hatch.
    """
    text = ""
    reasoning = ""
    tool_calls: List[ToolCallEvent] = []
    errors: List[ErrorEvent] = []
    end: Optional[RunEndEvent] = None

    for event in client.run(
        input=input,
        params=params,
        conversation_id=conversation_id,
        idempotency_key=idempotency_key,
    ):
        if isinstance(event, LlmDeltaEvent):
            text += event.text
            if on_text:
                on_text(event.text, event.nodeId)
        elif isinstance(event, ReasoningDeltaEvent):
            reasoning += event.text
            if on_reasoning:
                on_reasoning(event.text, event.nodeId)
        elif isinstance(event, ToolCallEvent):
            tool_calls.append(event)
            if on_tool_call:
                on_tool_call(event)
        elif isinstance(event, ErrorEvent):
            errors.append(event)
            if on_error:
                on_error(event)
        elif isinstance(event, RunEndEvent):
            end = event
        if on_event:
            on_event(event)

    if end is None:
        return StreamTextResult(
            text=text,
            reasoning=reasoning,
            runId="",
            status="failed",
            error=errors[-1].message if errors else "Stream ended without run.end.",
            durationMs=0,
            toolCalls=tool_calls,
            errors=errors,
        )

    return StreamTextResult(
        text=text,
        reasoning=reasoning,
        runId=end.runId,
        status=end.status,
        error=end.error,
        durationMs=end.durationMs,
        conversationId=end.conversationId,
        totalCostUsd=end.totalCostUsd,
        totalPromptTokens=end.totalPromptTokens,
        totalCompletionTokens=end.totalCompletionTokens,
        toolCalls=tool_calls,
        errors=errors,
    )


async def astream_text(
    client: AsyncPlumr,
    input: RunInput,
    *,
    params: Optional[Mapping[str, Any]] = None,
    conversation_id: Optional[str] = None,
    idempotency_key: Optional[str] = None,
    on_text: Optional[Callable[[str, str], None]] = None,
    on_reasoning: Optional[Callable[[str, str], None]] = None,
    on_tool_call: Optional[Callable[[ToolCallEvent], None]] = None,
    on_error: Optional[Callable[[ErrorEvent], None]] = None,
    on_event: Optional[Callable[[PlumrEvent], None]] = None,
) -> StreamTextResult:
    """Async variant of `stream_text`."""
    text = ""
    reasoning = ""
    tool_calls: List[ToolCallEvent] = []
    errors: List[ErrorEvent] = []
    end: Optional[RunEndEvent] = None

    async for event in client.run(
        input=input,
        params=params,
        conversation_id=conversation_id,
        idempotency_key=idempotency_key,
    ):
        if isinstance(event, LlmDeltaEvent):
            text += event.text
            if on_text:
                on_text(event.text, event.nodeId)
        elif isinstance(event, ReasoningDeltaEvent):
            reasoning += event.text
            if on_reasoning:
                on_reasoning(event.text, event.nodeId)
        elif isinstance(event, ToolCallEvent):
            tool_calls.append(event)
            if on_tool_call:
                on_tool_call(event)
        elif isinstance(event, ErrorEvent):
            errors.append(event)
            if on_error:
                on_error(event)
        elif isinstance(event, RunEndEvent):
            end = event
        if on_event:
            on_event(event)

    if end is None:
        return StreamTextResult(
            text=text,
            reasoning=reasoning,
            runId="",
            status="failed",
            error=errors[-1].message if errors else "Stream ended without run.end.",
            durationMs=0,
            toolCalls=tool_calls,
            errors=errors,
        )

    return StreamTextResult(
        text=text,
        reasoning=reasoning,
        runId=end.runId,
        status=end.status,
        error=end.error,
        durationMs=end.durationMs,
        conversationId=end.conversationId,
        totalCostUsd=end.totalCostUsd,
        totalPromptTokens=end.totalPromptTokens,
        totalCompletionTokens=end.totalCompletionTokens,
        toolCalls=tool_calls,
        errors=errors,
    )

File: typescript/src/streamText.ts
/**
 * `streamText` — high-level helper that mirrors the ergonomics of
 * AI-SDK's `streamText` while delegating to Plumr's event stream.
 *
 * Use this when you want callbacks (`onText`, `onToolCall`, ...) instead
 * of pattern-matching the raw event union.
 *
 * @example
 *   await streamText(plumr, {
 *     input: "Summarise the latest run",
 *     onText: (chunk) => process.stdout.write(chunk),
 *     onToolCall: ({ label }) => console.error(`→ ${label}`),
 *   });
 */

import type { Plumr } from "./client.js";
import type {
  ErrorEvent,
  PlumrEvent,
  ReasoningDeltaEvent,
  RunEndEvent,
  RunOptions,
  ToolCallEvent,
} from "./types.js";

export interface StreamTextOptions extends RunOptions {
  /** Called for every `llm.delta` (the orchestrator's token stream). */
  onText?: (text: string, ev: { nodeId: string }) => void;

  /** Called for every `reasoning.delta` (visible chain-of-thought). */
  onReasoning?: (text: string, ev: { nodeId: string }) => void;

  /** Called for every `tool.call`. */
  onToolCall?: (ev: ToolCallEvent) => void;

  /** Called for every typed `error` event from the runtime. */
  onError?: (ev: ErrorEvent) => void;

  /**
   * Called for every event, after the typed callback above (if any).
   * Useful as an escape hatch for events the helper doesn't surface
   * dedicated callbacks for (`step.start`, `step.end`, etc.).
   */
  onEvent?: (ev: PlumrEvent) => void;
}

export interface StreamTextResult {
  /** Concatenated `llm.delta` text from all orchestrator turns. */
  text: string;
  /** Concatenated `reasoning.delta` text. Empty when none was emitted. */
  reasoning: string;
  /** Final run id (from `run.end`). */
  runId: string;
  status: "succeeded" | "failed";
  error: string | null;
  durationMs: number;
  conversationId?: string;
  toolCalls: ToolCallEvent[];
  errors: ErrorEvent[];
}

export async function streamText(
  plumr: Plumr,
  opts: StreamTextOptions,
): Promise<StreamTextResult> {
  const {
    onText,
    onReasoning,
    onToolCall,
    onError,
    onEvent,
    ...runOpts
  } = opts;

  let text = "";
  let reasoning = "";
  const toolCalls: ToolCallEvent[] = [];
  const errors: ErrorEvent[] = [];
  let end: RunEndEvent | null = null;

  for await (const ev of plumr.run(runOpts)) {
    switch (ev.type) {
      case "llm.delta":
        text += ev.text;
        onText?.(ev.text, { nodeId: ev.nodeId });
        break;
      case "reasoning.delta": {
        const r = ev as ReasoningDeltaEvent;
        reasoning += r.text;
        onReasoning?.(r.text, { nodeId: r.nodeId });
        break;
      }
      case "tool.call":
        toolCalls.push(ev);
        onToolCall?.(ev);
        break;
      case "error":
        errors.push(ev);
        onError?.(ev);
        break;
      case "run.end":
        end = ev;
        break;
      // step.start / step.end / llm.start / llm.end / run.start /
      // run.cancelled fall through to the catch-all below.
      default:
        break;
    }
    onEvent?.(ev);
  }

  if (!end) {
    // Either the stream closed early (cancellation) or the runtime crashed.
    // Surface a synthetic failed result rather than throwing — callers can
    // still inspect `errors` / `toolCalls` / `text` collected up to that
    // point.
    return {
      text,
      reasoning,
      runId: "",
      status: "failed",
      error: errors[errors.length - 1]?.message ?? "Stream ended without run.end.",
      durationMs: 0,
      toolCalls,
      errors,
    };
  }

  return {
    text,
    reasoning,
    runId: end.runId,
    status: end.status,
    error: end.error,
    durationMs: end.durationMs,
    conversationId: end.conversationId,
    toolCalls,
    errors,
  };
}
